dojo/utils.py

```diff
@@ -2132,6 +2137,75 @@ def is_finding_groups_enabled():
     return get_system_setting("enable_finding_groups")
 
 
+class async_delete():
+    def __init__(self, *args, **kwargs):
+        self.mapping = {
+            'Product_Type': [
+                (Endpoint, 'product__prod_type'),
+                (Finding, 'test__engagement__product__prod_type'),
+                (Test, 'engagement__product__prod_type'),
+                (Engagement, 'product__prod_type'),
+                (Product, 'prod_type')],
+            'Product': [
+                (Endpoint, 'product'),
+                (Finding, 'test__engagement__product'),
+                (Test, 'engagement__product'),
+                (Engagement, 'product')],
+            'Engagement': [
+                (Finding, 'test__engagement'),
+                (Test, 'engagement')],
+            'Test': [(Finding, 'test')]
```

Comment:
Can you explain how this works? Are endpoints deleted when an Engagement is deleted, or can there be "orphaned" endpoints left around? What happens if a finding is deleted that is the original of a duplicate cluster?

Reply:
> Can you explain how this works? 

For each model in the dict, it will delete each model in the list in order. So for a product, it will delete endpoints -> findings -> tests -> engagements

> Are endpoints deleted when an Engagement is deleted, or can there be "orphaned" endpoints left around?

Endpoints have a foreign key relationship with products only. Findings have ManyToMany relationships with endpoints, so deleting any findings that have endpoints associated do not actually delete the endpoints. This removes the issue for any dangling objects.

> What happens if a finding is deleted that is the original of a duplicate cluster? Is there anything that needs to be done te remove any (dangling?) risk acceptances?

Duplicate clusters and resolving dangling ones is handled by the [findings delete method](https://github.com/DefectDojo/django-DefectDojo/blob/d07c007f24b1c6f0509b32e5aad3f2a2765b4581/dojo/finding/helper.py#L323). Risk Acceptances are deleted when an engagement is, and act as sort of a bucket for findings to be placed in. Deleting all findings associated with a risk acceptance should not delete the risk acceptance (which is current and expected behavior)